free_fleet/include/free_fleet/Manager.hpp

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#ifndef INCLUDE__FREE_FLEET__MANAGER_HPP
+#define INCLUDE__FREE_FLEET__MANAGER_HPP
+
+#include <memory>
+#include <vector>
+
+#include <rmf_utils/impl_ptr.hpp>
+#include <rmf_utils/optional.hpp>
+
+#include <rmf_traffic/Time.hpp>
+#include <rmf_traffic/agv/Graph.hpp>
+
+#include <free_fleet/agv/RobotInfo.hpp>
+#include <free_fleet/transport/Middleware.hpp>
+#include <free_fleet/CoordinateTransformer.hpp>
+
+#include <free_fleet/messages/Waypoint.hpp>
+#include <free_fleet/messages/ModeParameter.hpp>
+
+#include <free_fleet/messages/RobotState.hpp>
+#include <free_fleet/messages/ModeRequest.hpp>
+#include <free_fleet/messages/NavigationRequest.hpp>
+#include <free_fleet/messages/RelocalizationRequest.hpp>
+
+namespace free_fleet {
+
+class Manager
+{
+public:
+
+  using SharedPtr = std::shared_ptr<Manager>;
+
+  /// This function returns the a time stamp based on the implementation.
+  using TimeNow = std::function<rmf_traffic::Time()>;
+
+  /// This callback function can be used to work with different applications,
+  /// triggered every time a robot is updated with an incoming new state.
+  using RobotUpdatedCallback =
+    std::function<void(const std::shared_ptr<agv::RobotInfo>& robot_info)>;
+
+  /// Factory function that creates an instance of the Free Fleet Manager.
+  ///
+  /// \param[in] fleet_name
+  /// \param[in] graph
+  /// \param[in] middleware
+  /// \param[in] to_robot_transform
+  /// \param[in] time_now_fn
+  /// \param[in] robot_updated_callback_fn
+  /// \return
+  static SharedPtr make(
+    const std::string& fleet_name,
+    std::shared_ptr<rmf_traffic::agv::Graph> graph,
+    std::shared_ptr<transport::Middleware> middleware,
+    std::shared_ptr<CoordinateTransformer> to_robot_transform,
+    TimeNow time_now_fn,
+    RobotUpdatedCallback robot_updated_callback_fn);
+
+  /// Starts the manager which begins to listen for clients.
+  ///
+  /// \param[in] frequency
+  ///   Frequency at which the manager operates. This value needs to be a
+  ///   non-zero value.
+  void start(uint32_t frequency);
```

Comment:
The `start` function was added to the `Manager`since there was a reactive element to it, as well as with the idea of building applications around the `Manager`, I wanted to give users the option of starting the `Manager` only after everything is ready on the application (e.g. parameters, initialization, etc.)

Reply:
Based on our discussions, adding a `start_async(frequency)` and checking `started()` function, as well as mentioning in the API what they all do, https://github.com/open-rmf/free_fleet/pull/76/commits/b45c58b076e1dc07ff70065b5943cebbdf32fc26